src/api.rs

```diff
@@ -43,10 +44,12 @@ impl Api for GoApi {
             (self.vtable.canonicalize_address)(self.state, human_bytes, &mut output as *mut Buffer)
                 .into();
         let _human = unsafe { human_bytes.consume() };
-        let mut go_result: FfiResult<()> = go_result.into();
-        if let Err(ref mut error) = go_result {
-            error.set_message(format!("Failed to canonicalize the address: {}", human));
-        }
+        let go_result: FfiResult<()> = go_result.try_into().unwrap_or_else(|_| {
+            Err(FfiError::other(format!(
```

Comment:
A few points here that apply to all similar code:

1. This seems cleaner than the old approach - other must always have a message
2. The error information is only passed from go-cosmwasm to cosmwasm. I would need another (more go-centric) PR to bring the original error message from wasmd to cosmwasm, but at least the api to send it into the contract is there.
3. Why not merge the next line (`go_result?`) with this one (is it the FfiResult type declaration needed for `try_into()`? I still think there might be a cleaner way to express this, but definitely not a show stopper.

Reply:
Re 2: Yes, we now can add any string to `FfiError::Other`. However, I don't know where to get the strings from. The source `GoResult` ist just an enum/integer.

Re 3.: Because we implement the conversion `TryFrom<GoResult> for Result<(), FfiError>`, i.e. from an enum to a Result. We need the `let go_result: FfiResult<()>` part to derive the type for `try_into()`. Converting to the result seems to be (more than) enough for one expression and separating the success from the error case should be done in a different line I think.

Reply:
@webmaster128 @ethanfrey 
[This type definition in our repository](https://github.com/enigmampc/SecretNetwork/blob/757c43cc1e5dadfc3947bf9e764e51b83e6b666d/cosmwasm/lib/enclave-ffi-types/src/types.rs#L184) shows how you can pass rich enum types through FFI, but I think it is pretty difficult to represent rich enums on the Go side of things because it doesn't have support for union types. You could use a `*Buffer` argument with a call to `allocate_rust` for the error case, though, and on an error consume the argument and extract the error message from it.